app/helpers/random_assign_helper.rb

```diff
@@ -5,186 +5,234 @@ class UnableToRandomlyAssignGroupException < Exception
 
-      perform_group_exchange(reviewer, shuffle_index, next_index)
-      return
+  def attempt_assign_backwards(reviewer_id, shuffle_index)
+    (0...shuffle_index).each do |prev_index|
+      if can_assign_to_prev_index?(reviewer_id, prev_index, shuffle_index)
+        swap_previous_with_current_and_assign(reviewer_id, prev_index, shuffle_index)
+        return true
+      end
     end
 
-    raise UnableToRandomlyAssignGroupException
+    false
   end
 
-  def perform_group_exchange(reviewer, shuffle_index, prev_shuffle_index)
-    reviewee_at_shuffle_index = @shuffled_reviewees[shuffle_index]
-    reviewee_at_previous_index = @shuffled_reviewees[prev_shuffle_index]
-    peer_review = @shuffled_reviewees_pr[prev_shuffle_index]
-    reviewer_for_prev_reviewee = peer_review.reviewer
+  def create_peer_reviews_in_database
+    peer_reviews_reviewer_result = []
+    peer_reviews = []
+    results = []
+    marks = []
 
-    unlink_and_delete_previous_peer_review_assignment(peer_review, reviewee_at_previous_index)
-    swap_shuffled_indices(prev_shuffle_index, shuffle_index)
+    assignment_criteria = @pr_assignment.parent_assignment.get_criteria(:ta)
 
-    add_peer_review_to_db_and_remember_assignment(reviewer, reviewee_at_previous_index, shuffle_index)
-    add_peer_review_to_db_and_remember_assignment(reviewer_for_prev_reviewee, reviewee_at_shuffle_index, prev_shuffle_index)
-  end
+    # TODO - Mixing existing reviews causes nil elements in the list due to only needing to assign to a subset...
+    (0...@assigned_reviewer_to_shuffled_reviewee.size).each do |i|
+      reviewer_id = @assigned_reviewer_to_shuffled_reviewee[i]
+
+      # TODO - Remove this when nils are properly fixed (see above comment)
+      # NOTE: Only reason a nil value when we're completely and fully done assigning, so breaking is okay
+      # The final version obviously should not have this.
+      if reviewer_id.nil?
+        break
+      end
+
+      reviewee_id = @shuffled_reviewees[i]
+      reviewer = Grouping.find(reviewer_id)
+      reviewee = Grouping.find(reviewee_id)
+
+      result = Result.new(submission: reviewee.current_submission_used,
+                          marking_state: Result::MARKING_STATES[:incomplete])
+      results << result
+      assignment_criteria.each { |criterion| marks << criterion.marks.new(result: result) }
+      peer_reviews_reviewer_result << [reviewer, result]
+    end
+
+    Result.import results, :validate => false
+    Mark.import marks, :validate => false
+
+    # Peer reviews require IDs to have been made, so they come last.
+    peer_reviews_reviewer_result.each do |prdata|
+      peer_review = PeerReview.new(reviewer: prdata[0], result: prdata[1])
+      peer_reviews << peer_review
+    end
 
-  def unlink_and_delete_previous_peer_review_assignment(peer_review, reviewee)
-    @reviewers_assigned_to[peer_review.reviewer.id].delete(reviewee.id)
-    peer_review.destroy
+    PeerReview.import peer_reviews, :validate => false
```

Comment:
The good stuff is in the lines above. I think it looks a bit messy (since I'm making a pair out of an array, which I don't like doing as its cryptic, but this PR is to prove that it's working).